### PR TITLE
[SPIR-V] Support optimization of branching with analyzeBranch/removeBranch/insertBranch functions

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR fixes implementation of `SPIRVInstrInfo::analyzeBranch()` and adds implementations of `SPIRVInstrInfo::removeBranch()` and `SPIRVInstrInfo::insertBranch()` to support Branch Folding and If Conversion optimization.

The attached test case failed before this PR due to report_fatal_error() firing on missing implementation of `SPIRVInstrInfo::removeBranch()`.

The new test case is not able to pass spirv-val check at the moment due to the issue described in https://github.com/llvm/llvm-project/issues/110652 , this is not related to this PR.

This PR also updates instructions definition in tablegen to set isBranch=1 for relevant instructions.

### CODE REVIEW

**Comment by Keenuts** [comment]
> This PR fixes implementation of `SPIRVInstrInfo::analyzeBranch()` and adds implementations of `SPIRVInstrInfo::removeBranch()` and `SPIRVInstrInfo::insertBranch()` to support Branch Folding and If Conversion optimization.
> 
> The attached test case failed before this PR due to report_fatal_error() firing on missing implementation of `SPIRVInstrInfo::removeBranch()`.
> 
> The new test case is not able to pass spirv-val check at the moment due to the issue described in #110652 , this is not related to this PR.

I'd be careful about those.
The structurizer explicitly adds some empty BBs with a single `OpBranch` just to satisfy the SCFG constraints. Ex: a nested loop, with the inner loop with an early exit from the 2 loops. In such case, a block will be created just to the inner loop can break from 1 loop to the merge node, and then the second loop.
This branch will probably be deemed useless by any branch folding pass, but it is crucial to keep SPIR-V rules happy.

**Comment by VyacheslavLevytskyy** [comment]
> > This PR fixes implementation of `SPIRVInstrInfo::analyzeBranch()` and adds implementations of `SPIRVInstrInfo::removeBranch()` and `SPIRVInstrInfo::insertBranch()` to support Branch Folding and If Conversion optimization.
> > The attached test case failed before this PR due to report_fatal_error() firing on missing implementation of `SPIRVInstrInfo::removeBranch()`.
> > The new test case is not able to pass spirv-val check at the moment due to the issue described in #110652 , this is not related to this PR.
> 
> I'd be careful about those. The structurizer explicitly adds some empty BBs with a single `OpBranch` just to satisfy the SCFG constraints. Ex: a nested loop, with the inner loop with an early exit from the 2 loops. In such case, a block will be created just to the inner loop can break from 1 loop to the merge node, and then the second loop. This branch will probably be deemed useless by any branch folding pass, but it is crucial to keep SPIR-V rules happy.

Thank you! Let's try to harmonize those two parts then. Would it be possible to have several examples with such special cases, so that we are able to test how this PR changes behavior?

In general I'd say that this PR should not change anything to worse, because it implements a case that just failed before this PR. Changes in analyzeBranch() doesn't influence an optimization pass itself, but only make its logically consistent and conformant with the specification. In sum, where earlier report_fatal_error() was triggered, now we will have a transformation and succeeded translation, otherwise nothing changes. In the worst case, if your concerns materialized, we would get probably not so valid SPIRV code where previously we've always gotten crashes. I'd be anyway happy to work on this further and improve if you can help with examples where something can go wrong during optimization wrt. structurizer, instead of just crashing.

**Comment by VyacheslavLevytskyy** [comment]
@Keenuts Thank you for a good advice to be careful with analyzeBranch(). Testing and codegen source code helps to realize that the only safe option is to disable analyzeBranch(). One possible alternative to this would be to just entirely remove it from SPIRVInstrInfo, relying on default implementation. However, it seems better to keep it explicitly implemented with the comment as a way to document reasons to don't play with it in future. Please have a look at the result.

**Comment by Keenuts** [comment]
> @Keenuts Thank you for a good advice to be careful with analyzeBranch(). Testing and codegen source code helps to realize that the only safe option is to disable analyzeBranch(). One possible alternative to this would be to just entirely remove it from SPIRVInstrInfo, relying on default implementation. However, it seems better to keep it explicitly implemented with the comment as a way to document reasons to don't play with it in future. Please have a look at the result.

Thanks! I'm still dealing with structurizer bugs right now (found the reason for bad block sorting, and but fixing it triggers additional bugs). Sorry I hadn't had time to provide an example, but I'll bring some in the upcoming PR (simpler cases to showcase those things).

I think for the time being, not modifying branches after structuring might be for the best.
Thanks!

File: llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp [hunk 1]
```diff
@@ -212,9 +203,16 @@ bool SPIRVInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
 // If \p BytesRemoved is non-null, report the change in code size from the
 // removed instructions.
 unsigned SPIRVInstrInfo::removeBranch(MachineBasicBlock &MBB,
-                                      int *BytesRemoved) const {
-  report_fatal_error("Branch removal not supported, as MBB info not propagated"
-                     " to OpPhi instructions. Try using -O0 instead.");
+                                      int * /*BytesRemoved*/) const {
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp:206]: Should `BytesRemoved` be commented out?

File: llvm/test/CodeGen/SPIRV/branching/analyze-branch-opt.ll [hunk 2]
```diff
@@ -0,0 +1,250 @@
+; The goal of this test case is to ensure that translation does not crash when during branching
+; optimization analyzeBranch() requires helper methods of removeBranch() and insertBranch()
+; to manage subsequent operations.
+
+; RUN: llc -O0 -mtriple=spirv64-unknown-linux %s -o - | FileCheck %s
+; TODO: %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown %s -o - -filetype=obj | spirv-val %}
```
Comment (Keenuts) [inline llvm/test/CodeGen/SPIRV/branching/analyze-branch-opt.ll:6]: The TODO is here because of the block sorting issue no?

**Review by Keenuts — APPROVED** [review]

File: llvm/test/CodeGen/SPIRV/branching/analyze-branch-opt.ll (hunk 2 above)
Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/branching/analyze-branch-opt.ll:6]: yes, indeed

File: llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp:206]: It's not used (code size not handled), so I commented it out to avoid a warning.

**Review by michalpaszkowski — APPROVED** [review]

**Pull request merged**